### PR TITLE
[BEAM-8758] Google-cloud-spanner upgrade to 1.59.0 and google_cloud_bigtable_client_core to 1.16.0

### PR DESCRIPTION
**Pull request by terryxian78**
Fixes https://issues.apache.org/jira/browse/BEAM-8758
R: @chamikaramj 
CC: @suztomo
The changes are:

1. The main purpose of this commit is upgrading spanner library to 1.59.0, but to upgrade it, I need to also upgrade bigtable_client_core to 1.16.0 and other dependencies;
2. Update spanner type parser to be able to recognize NUMERIC.

Note: 
1. I have run "/bin/bash sdks/java/build-tools/beam-linkage-check.sh" and confirmed that there is no linkage errors.
2. I have run presubmit job and passed.
 ./gradlew spotlessApply && ./gradlew checkstyleMain checkstyleTest javadoc spotbugsMain compileJava compileTestJava

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace
--- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels
Python tests
Java tests

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by suztomo** [comment]
Failed: org.apache.beam.sdk.io.gcp.GcpApiSurfaceTest.testGcpApiSurface

The test ensures Beam does not expose unintended (internal) classes as Beam API.

**Comment by terryxian78** [comment]
> Failed: org.apache.beam.sdk.io.gcp.GcpApiSurfaceTest.testGcpApiSurface
> 
> The test ensures Beam does not expose unintended (internal) classes as Beam API.

The failure is https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/13264/testReport/org.apache.beam.sdk.io.gcp/GcpApiSurfaceTest/testGcpApiSurface/

The new spanner library needs to expose some classes in com.google.common.collect and interface com.google.common.base.Function, I will add these to the test.

**Comment by suztomo** [comment]
https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Commit/10003/ says:

```
Error Message
java.lang.RuntimeException: Workflow failed. Causes: The Compute Engine API has not fully initialized. Please try again in a
 few minutes., Unable to get machine type information for machine type n1-standard-4 in zone us-central1-a because of
 insufficient permissions. Please refer to https://cloud.google.com/dataflow/access-control#creating_jobs and make sure you
 have sufficient permissions.
```

**Comment by suztomo** [comment]
Run Java_Examples_Dataflow PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:174]: @terryxian78 Is this new?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java:68]: You may want to be more specific here. This includes all the classes in Guava's collection, ImmutableList, ImmutableSet, etc. I believe Beam team has intentionally avoided the classes to be exposed as APIs.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java
Comment (terryxian78) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:174]: No, This was included in my original pull request, I explained in the comment.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java
Comment (terryxian78) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java:68]: Ack, I will send another iteration to make it more specific.

Comment (terryxian78) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/GcpApiSurfaceTest.java]: Done!

**Comment by suztomo** [comment]
Run Python Precommit

**Comment by suztomo** [comment]
Run Typescript Precommit

**Review by nielm — APPROVED** [review]
LGTM
I am also curious why the new spanner library needs to expose Guava classes...

**Review by nielm — APPROVED** [review]
LGTM
Looks like the spanner API change that exposes Guava classes is:
[googleapis/java-spanner/pull/81](https://github.com/googleapis/java-spanner/pull/81),

Specifically:
[googleapis/java-spanner/pull/81/files#diff-7a9cb34faeb259be46b44f1878b7210f
](https://github.com/googleapis/java-spanner/pull/81/files#diff-7a9cb34faeb259be46b44f1878b7210f
)

**Comment by terryxian78** [comment]
Hi, @lukecwik, I was suggested by @suztomo to add you as reviewer. Could you review this change, or assign somebody else in committee to review it? Thanks!

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Comment by chamikaramj** [comment]
Run Dataflow ValidatesRunner

**Comment by chamikaramj** [comment]
Sorry, could you please address the open questions regarding API surface ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java
Comment (chamikaramj) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:174]: Can we move this to a separate PR with tests ?

**Comment by terryxian78** [comment]
> I can revisit when open questions are addressed. Thanks.

What open questions did you mean? All questions have been addressed, including  ApiSurfaceTest failure, and explanation of why new classes are exposed.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java
Comment (terryxian78) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java:174]: sure. I will make it a separate PR.

**Comment by chamikaramj** [comment]
Probably worth discussing the API surface updates in the mailing list before merging to confirm that there's no objection. Thanks.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java
Comment (terryxian78) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java]: Done! I reverted the change of this file.

**Comment by terryxian78** [comment]
> Run Java PostCommit
I am assuming that this comment will automatically kicks off postcommit test, I don't need to do anything, right? (this is my first PR, just want to make sure.)

**Comment by terryxian78** [comment]
Run Java Precommit

**Comment by terryxian78** [comment]
> Probably worth discussing the API surface updates in the mailing list before merging to confirm that there's no objection. Thanks.

You have asked whether there is objection in dev mailing list and I replied to it. Let me know what else I need to do and how long I should wait for objection? Thanks!

**Comment by chamikaramj** [comment]
Usual wait period for votes is three business days so we can do the same here.

BTW @nielm  mentioned that he create a bug in the Spanner side for updating the API. May be add a TODO with a link to that to remove new classes from the GcpApiSurfaceTest.java when/if that bug is fixed.

**Comment by terryxian78** [comment]
> Usual wait period for votes is three business days so we can do the same here.
> 
> BTW @nielm mentioned that he create a bug in the Spanner side for updating the API. May be add a TODO with a link to that to remove new classes from the GcpApiSurfaceTest.java when/if that bug is fixed.

@nielm, what is the bug number which @chamikaramj  talks about?

**Comment by nielm** [comment]
> @nielm, what is the bug number which @chamikaramj talks about?

[googleapis/java-spanner/issues/410](https://github.com/googleapis/java-spanner/issues/410)

There is a separate discussion in this bug that the suggested PR is not Java 1.7 compatible (due Future), and that it is a breaking change.

**Comment by terryxian78** [comment]
> > @nielm, what is the bug number which @chamikaramj talks about?
> 
> [googleapis/java-spanner/issues/410](https://github.com/googleapis/java-spanner/issues/410)
> 
> There is a separate discussion in that bug that the suggested PR is not Java 1.7 compatible (due Future), and that it is a breaking change.

Thanks, I have added a TODO including the link.

**Comment by terryxian78** [comment]
> Usual wait period for votes is three business days so we can do the same here.
> 
> BTW @nielm mentioned that he create a bug in the Spanner side for updating the API. May be add a TODO with a link to that to remove new classes from the GcpApiSurfaceTest.java when/if that bug is fixed.

Done, I have added a TODO including Niel's issue number.

**Comment by terryxian78** [comment]
> Usual wait period for votes is three business days so we can do the same here.
> 
> BTW @nielm mentioned that he create a bug in the Spanner side for updating the API. May be add a TODO with a link to that to remove new classes from the GcpApiSurfaceTest.java when/if that bug is fixed.

@chamikaramj It's been over 

> Usual wait period for votes is three business days so we can do the same here.
> 
> BTW @nielm mentioned that he create a bug in the Spanner side for updating the API. May be add a TODO with a link to that to remove new classes from the GcpApiSurfaceTest.java when/if that bug is fixed.

@chamikaramj , Three business days have passed, no objection in the dev mailing list. Are we good now? Thanks!

**Comment by terryxian78** [comment]
@chamikaramj, ping... 

I've resolved all concerns (no objection on GcpAPISurfaceTest), pls let me know if it is okay to merge? I have another PR which you asked me to do separately, so it would be nice to get this done so I can start that one.

**Review by chamikaramj — APPROVED** [review]
LGTM. Thanks.

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Comment by chamikaramj** [comment]
Run Dataflow ValidatesRunner

**Comment by chamikaramj** [comment]
Thanks. I'll merge this after post-commits pass.

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Pull request merged**
